src/procedures/proc_pagerank.c

```diff
@@ -0,0 +1,149 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "proc_pagerank.h"
+#include "../value.h"
+#include "../util/arr.h"
+#include "../query_ctx.h"
+#include "../util/rmalloc.h"
+#include "../graph/graphcontext.h"
+#include "../algorithms/pagerank.h"
+
+// CALL algo.pageRank('Page', 'LINKS', {iterations: 20, dampingFactor: 0.85}) YIELD node, score
+// CALL algo.pageRank('Page', 'LINKS') YIELD node, score
+
+typedef struct {
+	int n;                          // Number of nodes to rank.
+	int i;                          // Current node to return.
+	Graph *g;                       // Graph.
+	Node node;                      // Node.
+	GrB_Index *mappings;            // Mappings between extracted matrix rows and node ids.
+	LAGraph_PageRank *rankings;     // Nodes rankings.
+	SIValue *output;                // Array with 4 entries ["node", node, "score", score].
+} PagerankContext;
+
+ProcedureResult Proc_PagerankInvoke(ProcedureCtx *ctx, const char **args) {
+	if(array_len(args) != 2) return PROCEDURE_ERR;
+	const char *label = args[0];
+	const char *relation = args[1];
+
+	Schema *s = NULL;
+	GrB_Matrix l = NULL;
+	GrB_Matrix r = NULL;
+	Graph *g = QueryCtx_GetGraph();
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+
+	// Get label matrix.
+	s = GraphContext_GetSchema(gc, label, SCHEMA_NODE);
+	l = (s) ? Graph_GetLabelMatrix(g, s->id) : Graph_GetZeroMatrix(g);
+
+	// Get relation matrix.
+	s = GraphContext_GetSchema(gc, relation, SCHEMA_EDGE);
+	r = (s) ? Graph_GetRelationMatrix(g, s->id) : Graph_GetZeroMatrix(g);
+
+	// Mappings, array for returning row indices of tuples
+	GrB_Index n;
+	GrB_Matrix reduced;
+	GrB_Index *mappings = NULL;
+	GrB_Index rows = Graph_RequiredMatrixDim(g);
+	GrB_Index cols = rows;
+	assert(GrB_Matrix_nvals(&n, l) == GrB_SUCCESS);
+	assert(GrB_Matrix_new(&reduced, GrB_BOOL, n, n) == GrB_SUCCESS);
+
+	if(n != rows) {
+		mappings = rm_malloc(sizeof(GrB_Index) * n);
+		assert(GrB_Matrix_extractTuples_BOOL(mappings, GrB_NULL, GrB_NULL, &n, l) == GrB_SUCCESS);
```

Comment:
There's a memory leak here, but I'm not sure whether the fault is ours or not. Allocation:
```
192 bytes in 1 blocks are still reachable in loss record 168 of 225
   at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x9020A98: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
   by 0x902DB45: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
   by 0x9024EB5: GOMP_parallel (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
   by 0x8589929: GB_extract_vector_list (GB_extract_vector_list.c:68)
   by 0x858946A: GB_extractTuples (GB_extractTuples.c:115)
   by 0x856C669: GrB_Matrix_extractTuples_BOOL (GrB_Matrix_extractTuples.c:39)
   by 0x8554FD5: Proc_PagerankInvoke (proc_pagerank.c:58)
   by 0x85536C1: Proc_Invoke (procedure.c:67)
   by 0x84D34BC: OpProcCallInit (op_procedure_call.c:80)
   by 0x84C3CB3: _ExecutionPlanInit (execution_plan.c:952)
   by 0x84C3CEB: _ExecutionPlanInit (execution_plan.c:958)
```

Reply:
Can you please test `GrB_Matrix_extractTuples_BOOL` for leaks individually?
(outside of RedisGraph)